modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C

```diff
@@ -0,0 +1,271 @@
+  params.addParam<std::vector<Real>>("selfHardening_group_values",
+                                     "The self hardening coefficient q_aa"
+                                     "correspoinding to each group"
+                                     " i.e. '1.0 2.0 3.0' means 0-11 = 1.0, "
+                                     "12-23 = 2.0 and 24-48 = 3.0 "
+                                     " usually these are all 1.");
+  params.addParam<std::vector<Real>>("coplanarHardening_group_values",
+                                     "The coplanar laten hardening coefficient q_ab"
+                                     "correspoinding to each group"
+                                     " i.e. '1.0 2.0 3.0' means 0-11 = 1.0, "
+                                     "12-23 = 2.0 and 24-48 = 3.0 ");
+  params.addParam<std::vector<Real>>("GroupGroup_Hardening_group_values",
+                                     "The group-to-group laten hardening coefficient q_ab"
+                                     "This is a NxN vector"
+                                     " i.e. '1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0' "
+                                     "means non-coplanar slip systems in gr_11,22,33= "
+                                     "1.0, 5.0 and 9.0 respectively."
+                                     "latent ahrdening between for gr_12,13 = 2.0 3.0"
+                                     " respectively");
+  params.addClassDescription(
+      "Phenomenological Voce constitutive model state"
+      " variable evolution rate "
+      "component base class.  Override this virtual functions in your class");
+  return params;
+}
+
+CrystalPlasticityStateVarRateComponentVoce::CrystalPlasticityStateVarRateComponentVoce(
+    const InputParameters & parameters)
+  : CrystalPlasticityStateVarRateComponent(parameters),
+    _mat_prop_slip_rate(
+        getMaterialProperty<std::vector<Real>>(parameters.get<std::string>("uo_slip_rate_name"))),
+    _mat_prop_state_var(
+        getMaterialProperty<std::vector<Real>>(parameters.get<std::string>("uo_state_var_name"))),
+    _crystal_lattice_type(getParam<MooseEnum>("crystal_lattice_type")),
+    _groups(getParam<std::vector<unsigned int>>("groups")),
+    _h0_group_values(getParam<std::vector<Real>>("h0_group_values")),
+    _tau0_group_values(getParam<std::vector<Real>>("tau0_group_values")),
+    _tauSat_group_values(getParam<std::vector<Real>>("tauSat_group_values")),
+    _hardeningExponent_group_values(getParam<std::vector<Real>>("hardeningExponent_group_values")),
+    _selfHardening_group_values(getParam<std::vector<Real>>("selfHardening_group_values")),
+    _coplanarHardening_group_values(getParam<std::vector<Real>>("coplanarHardening_group_values")),
+    _GroupGroup_Hardening_group_values(
+        getParam<std::vector<Real>>("GroupGroup_Hardening_group_values"))
+{
+  // perform input checks and initialize usefull variables
+  _n_groups = _groups.size() - 1;
```

Comment:
Why `- 1`? This is an unusual convention. We use `_n_xxx` variables a lot and they usually correspond directly to the `size()` of an underlying container.

Reply:
By the way this is definitely wrong code. `_n_groups` is unsigned. So if `size()` is `0` it takes on a ginormous value (because it wraps around to the largest positive unsigned integer).

Reply:
Daniel, the variable _n_groups was set to be _groups.size() - 1 because in the input you definef group bounds. For example, let's say you are using BCC 24, and want to group slipsystems as primary and secondary then you will input 1 12 24. the format is 'format: [start end)'. You have 2 groups but 3 inputs. However I changed it to your standard notation. Now _n_groups = _groups.size().